docs(operator): restructure and improve developer instructions (#16326)
stackrox/stackrox: operator/DEFAULTING.md

# Defaulting

> This document provides more detail about possible ways to set defaults.

Reviewer: ```suggestion
This document provides more details on possible ways to set defaults.
```

You can safely skip it, if the simple instructions included in [EXTENDING_CRDs.md](EXTENDING_CRDS.md) satisfy your needs.

Generally speaking, the translation logic [described in the aforementioned document](EXTENDING_CRDS.md#map-crd-setting-to-helm-chart-configuration) will set the corresponding Helm values field only for explicitly set values.
For absent values, it will simply do nothing, thus deferring to the chart's defaulting logic (see [example](https://github.com/stackrox/rox/blob/84d841c870f59d2c423f78eb7ecd44a196f8a659/operator/pkg/central/values/translation/translation.go#L86)).

On one hand this allows some level of consistency, but on the other hand it is not great for maintainability,
because it requires diving into the Helm charts in order to discover what the default for the operator is.

> Therefore, it is better to explicitly provide a default for the operator.

Reviewer: ```suggestion
Therefore, it is usually preferable to explicitly provide defaults for the operator.
```

  value in the Helm values (see [example](https://github.com/stackrox/rox/blob/84d841c870f59d2c423f78eb7ecd44a196f8a659/operator/pkg/central/values/translation/translation.go#L120)).

  This method may be necessary in some cases (like mutually exclusive fields) but in general, if possible, please try to use the recommended one.

* **This mechanism is not used for the ACS operator.**

  Schema-level (a.k.a. static) defaults: These are set in the schema via `+kubebuilder` directives.
  If a field value is not set by the user, the default will be inserted automatically upon object creation and persisted.
  These values will be visible during translation, but only if the enclosing struct field is already present. Changing a schema-level default
  counts as a breaking API change, but it is treated as such only semantically, nothing will fail at runtime (see [example](https://github.com/stackrox/rox/blob/84d841c870f59d2c423f78eb7ecd44a196f8a659/operator/apis/platform/v1alpha1/central_types.go#L188))

  As of May 2025 we stopped adding such static defaults and are in the process of removing existing ones.
  Instead, use one of the above mechanisms.

## Defaulting Extension Mechanism

> The DefaultingExtension runs early in the reconcilliation process and executes "defaulting flows" in sequence.

Reviewer: ```suggestion
The DefaultingExtension runs early in the reconciliation process and executes "defaulting flows" in sequence.
```

Each defaulting flow has the ability to populate `Central.Defaults` (of type `CentralSpec`) resp. `SecuredCluster.Defaults` (of type `SecuredClusterSpec`).
These `.Defaults` fields are then applied onto their sibling `.Spec` fields in a way that:
- preserves user choices,
- does not persist in the cluster, such that we have the ability to dynamically change the default in the future.

### Generic static defaulting flow

There is a generic "static defaulting" flow, which is the appropriate place for defaulting of most simple cases.
To use it, simply add a default value for your new field to the `staticDefaults` struct and mention the default value in the last line of field description,
as [described in EXTENDING_CRDS.md](EXTENDING_CRDS.md).

### Custom defaulting flow

For more complex cases, such as using different defaults for upgrade vs. new installation scenarios, you should add a custom defaulting flow.
Such defaulting flow can:

> - Implement complex defaulting logic (beyond what static CRD defaulting supports).

Reviewer: ```suggestion
- Implement complex defaulting logic (beyond what static defaulting supports).
```

- Persist defaulting decisions in the custom resource's metadata as feature-specific annotation.
- Differentiate between green-field (fresh installation) and brown-field (upgrade) scenarios when making defaulting decision.

> - Ensure that subsequent reconciler extensions work with a custom resource spec that already includes all relevant defaulting decisions.

Reviewer: ```suggestion
- Ensure that subsequent reconciler extensions and the value translator work with a custom resource spec that already includes all relevant defaulting decisions.
```
